contracts/src/v0.8/automation/dev/test/AutomationRegistry2_3.t.sol

```diff
@@ -504,3 +505,100 @@ contract AutomationRegistry2_3_SetConfig is AutomationRegistry2_3_SetUp {
     return bytes32((prefix & prefixMask) | (h & ~prefixMask));
   }
 }
+
+contract AutomationRegistry2_3_NOPsSettlement is AutomationRegistry2_3_SetUp {
+  event NOPsSettledOffchain(address[] transmitterList, uint256[] balances);
+
+  function deployAndSetConfigForSettleOffchain(AutoBase2_3.PayoutMode payoutMode) public {
+    registryMaster = deployRegistry(payoutMode);
+    address module = address(new ChainModuleBase());
+    AutomationRegistryBase2_3.OnchainConfig memory cfg = AutomationRegistryBase2_3.OnchainConfig({
+      checkGasLimit: 5_000_000,
+      stalenessSeconds: 90_000,
+      gasCeilingMultiplier: 0,
+      maxPerformGas: 10_000_000,
+      maxCheckDataSize: 5_000,
+      maxPerformDataSize: 5_000,
+      maxRevertDataSize: 5_000,
+      fallbackGasPrice: 20_000_000_000,
+      fallbackLinkPrice: 2_000_000_000, // $20
+      fallbackNativePrice: 400_000_000_000, // $4,000
+      transcoder: 0xB1e66855FD67f6e85F0f0fA38cd6fBABdf00923c,
+      registrars: s_registrars,
+      upkeepPrivilegeManager: 0xD9c855F08A7e460691F41bBDDe6eC310bc0593D8,
+      chainModule: module,
+      reorgProtectionEnabled: true,
+      financeAdmin: FINANCE_ADMIN
+    });
+    bytes memory offchainConfigBytes = abi.encode(1234, ZERO_ADDRESS);
+
+    registryMaster.setConfigTypeSafe(
+      SIGNERS,
+      TRANSMITTERS,
+      F,
+      cfg,
+      OFFCHAIN_CONFIG_VERSION,
+      offchainConfigBytes,
+      new address[](0),
+      new AutomationRegistryBase2_3.BillingConfig[](0)
+    );
+  }
+
+  function testSettleNOPsOffchainRevertDueToUnauthorizedCaller() public {
+    deployAndSetConfigForSettleOffchain(AutoBase2_3.PayoutMode.ON_CHAIN);
+
+    vm.expectRevert(abi.encodeWithSelector(IAutomationRegistryMaster2_3.OnlyFinanceAdmin.selector));
+    registryMaster.settleNOPsOffchain();
+  }
+
+  function testSettleNOPsOffchainRevertDueToOffchainSettlementDisabled() public {
+    deployAndSetConfigForSettleOffchain(AutoBase2_3.PayoutMode.ON_CHAIN);
+
+    vm.prank(registryMaster.owner());
+    registryMaster.disableOffchainPayments();
+
+    vm.prank(FINANCE_ADMIN);
+    vm.expectRevert(abi.encodeWithSelector(IAutomationRegistryMaster2_3.MustSettleOnchain.selector));
+    registryMaster.settleNOPsOffchain();
+  }
+
+  function testSettleNOPsOffchainSuccess() public {
+    deployAndSetConfigForSettleOffchain(AutoBase2_3.PayoutMode.OFF_CHAIN);
+
+    uint256[] memory balances = new uint256[](TRANSMITTERS.length);
+    for (uint256 i = 0; i < TRANSMITTERS.length; i++) {
+      balances[i] = 0;
+    }
+
+    vm.startPrank(FINANCE_ADMIN);
+    vm.expectEmit();
+    emit NOPsSettledOffchain(TRANSMITTERS, balances);
+    registryMaster.settleNOPsOffchain();
+  }
```

Comment:
in this test, can we make some assertions that show the NOP balances have been reset to 0?

Reply:
this will require simulating a transmit. will work on it.